Userland/Applications/NetworkSettings/NetworkSettingsWidget.cpp

```diff
@@ -134,33 +139,72 @@ void NetworkSettingsWidget::on_switch_enabled_or_dhcp()
 
 void NetworkSettingsWidget::apply_settings()
 {
-    auto config_file = Core::ConfigFile::open_for_system("Network", Core::ConfigFile::AllowWriting::Yes).release_value_but_fixme_should_propagate_errors();
+    auto result = apply_settings_impl();
+    if (result.is_error()) {
+        GUI::MessageBox::show_error(window(), result.release_error().string_literal());
+        return;
+    }
+}
+
+ErrorOr<void> NetworkSettingsWidget::apply_settings_impl()
+{
+    auto json = TRY(create_settings_object());
+    if (json.is_empty())
+        return {};
+
+    auto pipefds = TRY(Core::System::pipe2(O_CLOEXEC));
+    ScopeGuard guard_fd1 { [&] { close(pipefds[1]); } };
+    {
+        posix_spawn_file_actions_t file_actions;
+        posix_spawn_file_actions_init(&file_actions);
+        posix_spawn_file_actions_adddup2(&file_actions, pipefds[0], STDIN_FILENO);
+
+        ScopeGuard guard_fd0_and_file_actions { [&]() {
+            posix_spawn_file_actions_destroy(&file_actions);
+            close(pipefds[0]);
+        } };
+
+        char const* argv[] = { "network-settings", nullptr };
+        (void)TRY(Core::System::posix_spawnp("network-settings"sv, &file_actions, nullptr, const_cast<char**>(argv), environ));
+
+        auto outfile = TRY(Core::File::adopt_fd(pipefds[1], Core::File::OpenMode::Write, Core::File::ShouldCloseFileDescriptor::No));
+        TRY(outfile->write_until_depleted(json.serialized<StringBuilder>().bytes()));
+    }
+
+    return {};
+}
+
+ErrorOr<JsonObject> NetworkSettingsWidget::create_settings_object()
+{
+    auto json = JsonObject();
     for (auto const& adapter_data : m_network_adapters) {
-        auto netmask = IPv4Address::netmask_from_cidr(adapter_data.value.cidr).to_deprecated_string();
-        config_file->write_bool_entry(adapter_data.key, "Enabled", adapter_data.value.enabled);
-        config_file->write_bool_entry(adapter_data.key, "DHCP", adapter_data.value.dhcp);
+        auto netmask = TRY(IPv4Address::netmask_from_cidr(adapter_data.value.cidr).to_string());
         if (adapter_data.value.enabled && !adapter_data.value.dhcp) {
             if (!IPv4Address::from_string(adapter_data.value.ip_address).has_value()) {
-                GUI::MessageBox::show_error(window(), DeprecatedString::formatted("Invalid IPv4 address for adapter {}", adapter_data.key));
-                return;
+                GUI::MessageBox::show_error(window(), TRY(String::formatted("Invalid IPv4 address for adapter {}", adapter_data.key)));
+                return JsonObject();
```

Comment:
Why are you returning a valid object if you return early due to an error?
`return Error::from_string_literal("Invalid IPv4 address")`?

Reply:
Because I dont want to show the error message twice?

Reply:
I replaced it with an `Optional<JsonObject> {}`